ast/check.go

```diff
@@ -61,124 +81,211 @@ func (tc *typeChecker) CheckBody(env *TypeEnv, body Body) (*TypeEnv, Errors) {
 	return env, tc.errs
 }
 
-// CheckRules runs type checking on the rules and returns a TypeEnv if no
-// errors are found. The resulting TypeEnv wraps the provided one. The
-// resulting TypeEnv will be able to resolve types of refs that refer to rules.
-func (tc *typeChecker) CheckRules(env *TypeEnv, rules []*Rule) (*TypeEnv, Errors) {
+func (tc *typeChecker) checkLanguageBuiltins(env *TypeEnv) {
+	for _, bi := range Builtins {
```

Comment:
checkLanguageBuiltins, checkFunc, and checkRule are all inserted type info into `env`.

In other words, checkLanguageBuiltins could be done outside of CheckTypes. This is a minor point, we can come back and refactor later on. Consider this resolved.